source/comp/markv_codec.cpp

```diff
@@ -0,0 +1,1194 @@
+                                    model_->s16_block_exponent());
+    } else if (operand.number_kind == SPV_NUMBER_FLOATING) {
+      writer_.WriteUnencoded(word);
+    } else {
+      assert(0);
+    }
+  } else {
+    assert(operand.number_bit_width == 64);
+    const uint64_t word =
+        uint64_t(instruction.word(operand.offset)) |
+        (uint64_t(instruction.word(operand.offset + 1)) << 32);
+    if (operand.number_kind == SPV_NUMBER_UNSIGNED_INT) {
+      writer_.WriteVariableWidthU64(word, model_->u64_chunk_length());
+    } else if (operand.number_kind == SPV_NUMBER_SIGNED_INT) {
+      int64_t val = 0;
+      memcpy(&val, &word, 8);
+      writer_.WriteVariableWidthS64(val, model_->s64_chunk_length(),
+                                    model_->s64_block_exponent());
+    } else if (operand.number_kind == SPV_NUMBER_FLOATING) {
+      writer_.WriteUnencoded(word);
+    } else {
+      assert(0);
+    }
+  }
+}
+
+spv_result_t MarkvEncoder::EncodeInstruction(
+    const spv_parsed_instruction_t& inst) {
+  const spv_result_t validation_result = UpdateValidationState(inst);
+  if (validation_result != SPV_SUCCESS)
+    return validation_result;
+
+  assert(vstate_.unresolved_forward_id_count() == 0 &&
+         "Forward declared id support not yer implemented");
+
+  const Instruction& instruction = GetCurrentInstruction();
+  const auto& operands = instruction.operands();
+
+  LogDisassemblyInstruction();
+
+  // Write opcode.
+  writer_.WriteVariableWidthU16(inst.opcode, model_->opcode_chunk_length());
+
+  if (OpcodeHasVariableNumberOfOperands(SpvOp(inst.opcode))) {
+    // Write num_operands.
+    if (logger_)
+      logger_->AppendWhitespaces(kCommentNumWhitespaces);
+
+    writer_.WriteVariableWidthU16(inst.num_operands,
+                                  model_->num_operands_chunk_length());
+  }
+
+  // Write operands.
+  for (const auto& operand : operands) {
+    if (operand.type == SPV_OPERAND_TYPE_RESULT_ID) {
+      // Result ids are not written (will be regenerated by the decoder).
+      // TODO: Support forward declared ids.
+      continue;
+    }
+
+    if (logger_)
+      logger_->AppendWhitespaces(kCommentNumWhitespaces);
+
+    if (operand.type == SPV_OPERAND_TYPE_TYPED_LITERAL_NUMBER) {
+      EncodeLiteralNumber(instruction, operand);
+    } else if (operand.type == SPV_OPERAND_TYPE_LITERAL_STRING) {
+      const char* src =
+          reinterpret_cast<const char*>(&instruction.words()[operand.offset]);
+      const size_t length = spv_strnlen_s(src, operand.num_words * 4);
+      if (length == operand.num_words * 4)
+        return vstate_.diag(SPV_ERROR_INVALID_BINARY)
+            << "Failed to find terminal character of literal string";
+      for (size_t i = 0; i < length + 1; ++i)
+        writer_.WriteUnencoded(src[i]);
+    } else {
+      for (int i = 0; i < operand.num_words; ++i) {
+        const uint32_t word = instruction.word(operand.offset + i);
+        EncodeOperandWord(operand.type, word);
```

Comment:
It's not so much a bug, but a question if I want to run
StripDebugInfoPass
FlattenDecorationPass
CompactIdsPass
on all binaries coming to the encoder.